server/portal/apps/onboarding/steps/project_membership.py

```diff
@@ -1,69 +1,243 @@
-                baseURL=settings.TAS_URL,
-                credentials={
-                    'username': settings.TAS_CLIENT_KEY,
-                    'password': settings.TAS_CLIENT_SECRET,
-                }
+            tas_client.add_project_user(self.settings['project_sql_id'], self.user.username)
+        except Exception as e:
+            error, reason = e.args
+            if "is already a member" in reason:
+                self.complete(
+                    "{username} is already a member of the {project}".format(
+                        username=self.user.username,
+                        project=self.project['title']
+                    )
+                )
+            else:
+                self.fail(
+                    "{username} could not be added to {project} due to error {reason}".format(
+                        project=self.project['title'],
+                        username=self.user.username,
+                        reason=reason
+                    )
+                )
+                raise e
+
+    def deny_project_request(self):
+        ticket_id = None
+        for event in self.events:
+            if event.data and "ticket" in event.data:
+                ticket_id = event.data["ticket"]
+        tracker = self.get_tracker()
+        request_text = """Your request for membership on the {project} project has been
+        denied. If you believe this is an error, please submit a help ticket.
+        """.format(
+            project=self.project['title']
+        )
+        if tracker.login():
+            tracker.reply(ticket_id, text=request_text)
+            tracker.comment(
+                ticket_id,
+                text="User was not added to the {project} TAS Project (GID {gid}) at {base_url}".format(
+                    project=self.project['title'],
+                    gid=self.project['gid'],
+                    base_url=settings.WH_BASE_URL
+                )
             )
-        except Exception:
+            tracker.edit_ticket(ticket_id, Status='resolved')
+        else:
             self.fail(
-                "Unable to get tas in order to get project info")
-            return
+                "The portal was unable to close RT Ticket {ticket_id}".format(
+                    ticket_id=ticket_id
+                )
+            )
 
-        try:
-            projects = tas_client.projects_for_user(username=self.user.username)
+    def close_project_request(self, deny=False):
+        ticket_id = None
+        for event in self.events:
+            if event.data and "ticket" in event.data:
+                ticket_id = event.data["ticket"]
+        tracker = self.get_tracker()
+        request_text = """Your request for membership on the {project} project has been
+        granted. Please login at {base_url}/onboarding/setup to continue setting up your account.
+        """.format(
+            project=self.project['title'],
+            base_url=settings.WH_BASE_URL
```

Comment:
Might be out of scope of this PR at this point, but is it possible to use something other than `WH_BASE_URL` here? That setting's original intent was to allow local debugging with an ngrok url

Reply:
TLDR; WH_BASE_URL is now a required setting in production deployments of CEPv1 for many reasons.

The previous way of generating a callback URL is by deriving it using [django's request build_absolute_uri](https://github.com/django/django/blob/master/django/http/request.py). This has several implications:

- Explicit is more convenient than dyanmic. Anything requiring a callback must have a client request as the source, and that data must be passed back and forth. It's inconvenient, and in the case of doing anything involving the python management console and onboarding steps it is difficult to simulate. For example, last week we ran a2cps's SystemCreation step manually on staging in the python shell to add secure_data for everyone using the console. If we had to rely on a request object, this would have needed more complex code to generate a request object and i'm not sure it's possible to inject the headers necessary to specify an origin in the python shell. Even if we could generate the request objects, this requires more parameter passing (either the request object or the string containing the dynamically generated base URL)
- Explicit is more secure than dynamic. build_absolute_uri derives the domain from the incoming request headers from the client. In the (unlikely, but we never use "unlikely" as a synonym for "secure") situation where a request header is falsified by a client agent and the origin points to a third party, we are now passing this third party origin as the webhook callback destination to TAPIS and other code, which may send information to the third party. In the case of this particular link generating code, we are now asking TACC staff to click on a link that could potentially take them to a malicious site. Therefore, my bones tell me that an explicitly specified webhook callback URL is more secure.
- There's now a webhook callback mechanism that is currently only being used by the system creation application. While we have code that receives job status updates at a webhook that does not use it, it's cumbersome. We have had to patch that end point last year to verify the caller origin (which is why i'm asking CIC to implement a feature in TAPIS 3 to sign those webhooks somehow). The code to verify the incoming job post data is replicated in the handler for /webhooks/interactive. For code reuse purposes, I think it's important to eventually migrate to the new webhook callback generator with nonce and then if we want to settle the question of WH_BASE_URL changes should be made there.

That being said, I acknowledge that this creates a bug during testing of this step. RT tickets generated from local test environments will have the ngrok url rather than cep.dev. I had a long discussion about how to fix this with @nathanfranklin and whether or not it would be worth it. My opinion is that nobody has been looking at the generated consult tickets from local testing as part of the validation of this step, and we won't get any complaints. The alternative is to introduce more code branching and more settings_secret values.

Reply:
Good points, fair enough. Agree on the test tickets, not important